File: FDM/FDM_1D.py
import numpy as np # type: ignore
import matplotlib.pyplot as plt # type: ignore

# Parameters
dx = 1
dt = 0.01
x_0 = float(input("Enter Initial Value of X , x_0: "))
x_Nx = float(input("Enter Final value of X, x_Nx: "))
t_0 = float(input("Enter Initial Value of t , t_0: "))
t_Nt = float(input("Enter Final value of t, t_Nt: "))

Nt = round((t_Nt - t_0) / dt)
Nx = round((x_Nx - x_0) / dx)

print("Total No. of discrete points of time t:", Nt + 1)
print("Total No. of discrete points of x:", Nx + 1)

k = 1.0
F = (k * dt) / (dx * dx)
print("F =", F)

# Initialize temperature matrix: Nt+1 time steps and Nx+1 space points
T = np.full((Nt + 1, Nx + 1), 27.0)

# Boundary conditions
T[:, 0] = 200.0
T[:, Nx] = 500.0

# Time stepping
for n in range(Nt):
    for i in range(1, Nx):
        T[n + 1][i] = F * T[n][i - 1] + (1 - 2 * F) * T[n][i] + F * T[n][i + 1]

# Output
print("\nTime step results (T[n][i]):")
for n in range(Nt + 1):
    print(f"t = {n * dt:.2f}s: ", end='')
    for i in range(Nx + 1):
        print(f"{T[n][i]:.2f}", end=' ')
    print()


x = np.linspace(x_0, x_Nx, Nx + 1)
time_indices = np.linspace(0, Nt, 6, dtype=int)  # 6 equally spaced time steps

plt.figure(figsize=(10, 6))
for n in time_indices:
    plt.plot(x, T[n], label=f"t = {n * dt:.2f} s")

plt.title("1D Heat Equation (Explicit Scheme)")
plt.xlabel("x")
plt.ylabel("Temperature (°C)")
plt.legend()
plt.grid(True)
plt.show()

# Generate x and t arrays for labeling
x = np.linspace(x_0, x_Nx, Nx + 1)
t = np.linspace(t_0, t_Nt, Nt + 1)

# Plot 2D heatmap
plt.figure(figsize=(12, 6))
plt.imshow(T, aspect='auto', extent=[x_0, x_Nx, t_Nt, t_0], cmap='hot')
plt.colorbar(label="Temperature (°C)")
plt.title("Temperature Distribution over Space and Time")
plt.xlabel("Position x")
plt.ylabel("Time t (s)")
plt.tight_layout()
plt.show()

File: FDM/FDM_1D.cpp
#include <iostream>
#include <vector>
#include <cmath>
using namespace std;

int main(){
    double x_0, x_Nx, t_0, t_Nt;
    double dx = 1, dt = 0.01;

    cout<<"Enter Initial Value of X , x_0 : ";
    cin>>x_0;

    cout<<"Enter Final value of X, x_Nx : ";
    cin>>x_Nx;

    cout<<"Enter Initial Value of t , t_0 : ";
    cin>>t_0;

    cout<<"Enter Final value of t, t_Nt : ";
    cin>>t_Nt;

    // dT/dx = k * d2T/dx2
    //T(x, 0) = 27 C
    //T(x_0, t) = 200 C
    //T(x_N, t) = 500 C
    //0 <= t <= 1
    //0 <= x <= 4

    int Nt;
    Nt = round((t_Nt - t_0) / dt);
    cout<<"Total No. of descrete points of time t : "<<Nt + 1<<endl;

    int Nx ;
    Nx = (x_Nx - x_0)/dx;
    cout<<"Total No. of descrete points of x : "<<Nx + 1<<endl;

    double k = 1.0;
    double F;
    F = (k * dt)/(dx * dx);

    cout<<"F = "<<F<<endl;

    vector<vector<double>> T(Nt + 1, vector<double>(Nx + 1, 27.0));


    for (int n = 0; n <= Nt; n++){
        T[n][0] = 200.0;
        T[n][Nx] = 500.0;
    }

    
    for (int n = 0; n < Nt; n++){
        for(int i = 1 ; i < Nx; i++){
            T[n + 1][i] = F * T[n][i-1] + (1-2*F) * T[n][i] + F * T[n][i+1];
        }
    }


    cout << "Time step results (T[i][j]):\n";
    for (int n = 0; n <= Nt; n++) {
        cout << "t = " << n * dt << "s: ";
        for (int i = 0; i <= Nx; i++) {
            cout << T[n][i] << " ";
        }
        cout << endl;
    }


    return 0;
}